python/hail/typecheck/check.py

```diff
@@ -213,67 +330,103 @@ def check_all(f, args, kwargs, checks, is_method):
     spec = getargspec(f)
     name = f.__name__
 
-    arg_names = []
-    args_to_check = []
+    args_ = []
 
     # strip the first argument if is_method is true (this is the self parameter)
     if is_method:
         if not (len(args) > 0 and isinstance(args[0], object)):
             raise RuntimeError(
                 '%s: no class found as first argument. Use typecheck instead of typecheck_method?' % name)
-        arg_names.extend(spec.args[1:])
-        args_to_check.extend(args[1:])
+        named_args = spec.args[1:]
+        pos_args = args[1:]
+        args_.append(args[0])
     else:
-        arg_names.extend(spec.args)
-        args_to_check.extend(args)
+        named_args = spec.args[:]
+        pos_args = args[:]
 
-    # if f has varargs, tuple any unnamed args and pass that as a regular argument to the checker
-    if spec.varargs:
-        n_named_args = len(spec.args) - (1 if is_method else 0)
-        tupled_varargs = tuple(args_to_check[n_named_args:])
-        args_to_check = args_to_check[:n_named_args]
-        args_to_check.append(tupled_varargs)
-        arg_names.append(spec.varargs)
-
-    # if f has varkw, pass them as a dict to the checker.
-    if spec.varkw:
-        args_to_check.append(kwargs)
-        arg_names.append(spec.varkw)
+    signature_namespace = set(named_args).union(
+        set(filter(lambda x: x is not None, [spec.varargs, spec.varkw])))
+    tc_namespace = set(checks.keys())
 
     # ensure that the typecheck signature is appropriate and matches the function signature
-    if set(arg_names) != set(checks.keys()):
-        unmatched_tc = [k for k in checks if k not in arg_names]
-        unmatched_f = [k for k in arg_names if k not in checks]
-        if unmatched_f or unmatched_tc:
+    if signature_namespace != tc_namespace:
+        unmatched_tc = list(tc_namespace - signature_namespace)
+        unmatched_sig = list(signature_namespace - tc_namespace)
+        if unmatched_sig or unmatched_tc:
             msg = ''
             if unmatched_tc:
                 msg += 'unmatched typecheck arguments: %s' % unmatched_tc
-            if unmatched_f:
+            if unmatched_sig:
                 if msg:
                     msg += ', and '
-                msg += 'function parameters with no defined type: %s' % unmatched_f
+                msg += 'function parameters with no defined type: %s' % unmatched_sig
             raise RuntimeError('%s: invalid typecheck signature: %s' % (name, msg))
 
-    # type check the function arguments
-    for argname, arg in zip(arg_names, args_to_check):
-        tc = only(checks[argname])
-
-        if not tc.check(arg):
-            raise TypeError("%s: parameter '%s': expected %s, found %s: '%s'" %
-                            (name, argname, tc.expects(), extract(type(arg)), str(arg)))
+    # if f has varargs, tuple any unnamed args and pass that as a regular argument to the checker
```

Comment:
this confused me since you aren't the one doing the tupling. how about: `# if f has a vararg parameter, pass this corresponding tuple as a regular argument to the checker`

Reply:
just deleted the comment, it was out of date.